monkey/monkey_island/cc/agent_event_handlers/scan_event_handler.py

```diff
@@ -55,71 +62,54 @@ def handle_tcp_scan_event(self, event: TCPScanEvent):
             return
 
         try:
-            target_machine = self._get_target_machine(event)
-            source_node = self._get_source_node(event)
+            tcp_connections = self._get_tcp_connections_from_event(event)
+            network_services = self._get_network_services_from_event(event)
 
-            self._update_nodes(target_machine, event)
-            self._update_tcp_connections(source_node, target_machine, event)
-            self._update_network_services(target_machine, event)
+            self._upsert_from_tcp_scan_event(event, tcp_connections, network_services)
         except (RetrievalError, StorageError, UnknownRecordError):
             logger.exception("Unable to process tcp scan data")
 
-    def _get_target_machine(self, event: ScanEvent) -> Machine:
-        try:
-            target_machines = self._machine_repository.get_machines_by_ip(event.target)
-            return target_machines[0]
-        except UnknownRecordError:
-            machine = Machine(
-                id=self._machine_repository.get_new_id(),
-                network_interfaces=[IPv4Interface(event.target)],
-            )
-            self._machine_repository.upsert_machine(machine)
-            return machine
-
-    def _get_source_node(self, event: ScanEvent) -> Node:
-        machine = self._get_source_machine(event)
-        try:
-            node = self._node_repository.get_node_by_machine_id(machine.id)
-        except UnknownRecordError:
-            node = Node(machine_id=machine.id)
-            self._node_repository.upsert_node(node)
-        return node
-
-    def _get_source_machine(self, event: ScanEvent) -> Machine:
-        agent = self._agent_repository.get_agent_by_id(event.source)
-        return self._machine_repository.get_machine_by_id(agent.machine_id)
+    @staticmethod
+    def _get_open_ports(event: TCPScanEvent) -> List[NetworkPort]:
+        return [port for port, status in event.ports.items() if status == PortStatus.OPEN]
 
     def _update_target_machine_os(self, machine: Machine, event: PingScanEvent):
         if event.os is not None and machine.operating_system is None:
             machine.operating_system = event.os
             self._machine_repository.upsert_machine(machine)
 
-    def _update_network_services(self, target: Machine, event: TCPScanEvent):
-        network_services = {
+    @classmethod
```

Comment:
Probably no point in this being a class method. We could do `ScanEventHandler._get_open_ports(event)`

Reply:
I researched and I think it's OK to use cls, since it calls `_get_open_ports`. I've not found any proper arguments in favor of one or the other. Static methods hint that this method doesn't need to access any state (which it doesn't), but classmethod handles inheritance better because it allows us to override the `_get_open_ports` in a child and this method will still work(which we also don't care about in this case)